File: kai-printers-android/app/src/main/java/com/kaistore/printers/print/CashDrawerPolicy.kt
package com.kaistore.printers.print

/**
 * Política de gaveta alineada con Kai Printers desktop (`cash_drawer_policy.rs`)
 * y `@kai/print-service-client` (`cash-drawer-policy.ts`).
 */
object CashDrawerPolicy {
    private val ELIGIBLE_TYPES = setOf(
        "pos-sale-ticket",
        "fiscal-boleta-preview",
        "pos-cash-session-opening-ticket",
        "pos-cash-count-sheet-ticket",
        "pos-cash-hub-movement-ticket",
        "pos-supplier-payment-ticket",
        "pos-payment-in-ticket",
        "test_print",
        "test_escpos_qa",
        "test_escpos_qa_nocut",
    )

    fun agentTypeMayOpenDrawer(agentType: String?): Boolean {
        val t = agentType?.trim().orEmpty()
        if (t.isEmpty()) return false
        if (t == "test_drawer") return true
        return ELIGIBLE_TYPES.contains(t)
    }

    /**
     * @param drawerEnabledInMapping switch «Apertura de gaveta» de la línea Tickets (si no hay mapeo, false).
     */
    fun shouldOpenDrawer(
        agentType: String?,
        widthChars: Int,
        drawerEnabledInMapping: Boolean = false,
    ): Boolean {
        if (!EscPosTail.shouldOpenCashDrawer(widthChars)) return false
        val t = agentType?.trim().orEmpty()
        if (t == "test_drawer") return true
        if (!drawerEnabledInMapping) return false
        return agentTypeMayOpenDrawer(t)
    }
}

File: packages/print-service-client/src/cash-drawer-policy.ts
/**
 * Tipos agente (`type` del protocolo print) que pueden abrir gaveta en 80 mm
 * si la línea Tickets tiene «Apertura de gaveta» activa en Kai Printers.
 *
 * Debe mantenerse alineado con `cash_drawer_policy.rs` (desktop) y `CashDrawerPolicy.kt` (Android).
 */
export const CASH_DRAWER_ELIGIBLE_AGENT_TYPES = new Set([
  "pos-sale-ticket",
  "fiscal-boleta-preview",
  "pos-cash-session-opening-ticket",
  "pos-cash-count-sheet-ticket",
  "pos-cash-hub-movement-ticket",
  "pos-supplier-payment-ticket",
  "pos-payment-in-ticket",
  "test_print",
  "test_escpos_qa",
  "test_escpos_qa_nocut",
]);

/** Pulso de gaveta en prueba dedicada (ignora switch de mapeo en desktop). */
export const CASH_DRAWER_TEST_AGENT_TYPE = "test_drawer";

export function agentPrintTypeMayOpenCashDrawer(agentType: string | null | undefined): boolean {
  const t = (agentType ?? "").trim();
  if (!t) return false;
  if (t === CASH_DRAWER_TEST_AGENT_TYPE) return true;
  return CASH_DRAWER_ELIGIBLE_AGENT_TYPES.has(t);
}

/** Ancho mínimo de ticket (columnas) para gaveta: 48 = 80 mm, 32 = 58 mm. */
export const CASH_DRAWER_MIN_WIDTH_CHARS = 48;

export function ticketWidthCharsMayOpenCashDrawer(widthChars: number): boolean {
  return widthChars >= CASH_DRAWER_MIN_WIDTH_CHARS;
}

export function shouldOpenCashDrawerForTicketJob(
  agentType: string | null | undefined,
  widthChars: number,
  drawerEnabledInMapping: boolean,
): boolean {
  if (!ticketWidthCharsMayOpenCashDrawer(widthChars)) return false;
  const t = (agentType ?? "").trim();
  if (t === CASH_DRAWER_TEST_AGENT_TYPE) return true;
  if (!drawerEnabledInMapping) return false;
  return agentPrintTypeMayOpenCashDrawer(t);
}
